Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -0,0 +1,403 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidAlgorithms/MultipleScatteringCorrection.h"
+#include "MantidAPI/HistoWorkspace.h"
+#include "MantidAPI/InstrumentValidator.h"
+#include "MantidAPI/MatrixWorkspace_fwd.h"
+#include "MantidAPI/Sample.h"
+#include "MantidAPI/SampleValidator.h"
+#include "MantidAPI/SpectrumInfo.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidAPI/WorkspaceUnitValidator.h"
+#include "MantidAlgorithms/MultipleScatteringCorrectionDistGraber.h"
+#include "MantidDataObjects/WorkspaceCreation.h"
+#include "MantidGeometry/IDetector.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidGeometry/Instrument/SampleEnvironment.h"
+#include "MantidHistogramData/Interpolate.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+
+namespace Mantid {
+
+namespace Algorithms {
+
+using namespace API;
+using namespace Geometry;
+using namespace Kernel;
+using namespace Mantid::DataObjects;
+
+using HistogramData::interpolateLinearInplace;
+
+DECLARE_ALGORITHM(MultipleScatteringCorrection)
+
+namespace {
+// the maximum number of elements to combine at once in the pairwise summation
+constexpr size_t MAX_INTEGRATION_LENGTH{1000};
+
+inline size_t findMiddle(const size_t start, const size_t stop) {
+  auto half = static_cast<size_t>(floor(.5 * (static_cast<double>(stop - start))));
+  return start + half;
+}
+
+inline size_t calcLinearIdxFromUpperTriangular(const size_t N, const size_t row_idx, const size_t col_idx) {
+  // calculate the linear index from the upper triangular matrix from a (N x N) matrix
+  // row_idx < col_idx due to upper triangular matrix
```

Comment:
Would it be good to add an `assert(row_idx < col_idx)` here? That would reiterate the assumption in the comment, and not be run except for debug builds. This will help in the future as we expand functionality to help track down issues discovered by tests. It would be even better if it could be a [`static_assert`](https://en.cppreference.com/w/cpp/language/static_assert), but these values likely aren't known at compile-time.

Reply:
the recommended assertion statement has been added